api/src/services/users.ts

```diff
@@ -451,6 +458,111 @@ export class UsersService extends ItemsService {
 		await service.updateOne(user.id, { password, status: 'active' });
 	}
 
+	async registerUser(email: string, password: string) {
+		const STALL_TIME = 750;
+		const timeStart = performance.now();
+		const serviceOptions: AbstractServiceOptions = { accountability: this.accountability, schema: this.schema };
+		const settingsService = new SettingsService(serviceOptions);
+
+		const settings = await settingsService.readSingleton({
+			fields: [
+				'is_public_registration_enabled',
+				'is_public_registration_email_validation_enabled',
+				'public_registration_role',
+				'public_registration_email_filter',
+			],
+		});
+
+		if (settings?.['is_public_registration_enabled'] == false) {
+			await stall(STALL_TIME, timeStart);
+			throw new ForbiddenError();
+		}
+
+		const publicRegistrationRole = settings?.['public_registration_role'];
+
+		if (!publicRegistrationRole) {
+			await stall(STALL_TIME, timeStart);
+			throw new ContainsNullValuesError({ collection: 'directus_settings', field: 'public_registration_role' });
+		}
+
+		const hasEmailValidation = settings?.['is_public_registration_email_validation_enabled'];
+
+		const partialUser: Partial<User> = {
+			email,
+			password,
+			role: publicRegistrationRole,
+			status: hasEmailValidation ? 'draft' : 'active', // TODO: Do we want to have a dedicated "unverified" status?
```

Comment:
I'm thinking about renaming the `invited`-status to a more generic `pending`-status so that we can use the same status for invited and registered-but-not-yet-verified users but simply using the migration file to rename the status could lead to breaking existing logic in users projects for example somebody might have Flows that expect the users to be `invited`

Another approach could be to just replace the `Invited` UI label to `Pending` and keep the underlying `invited` string the same to stay backwards compatible. 

Mhh. Could also just add a new status to be both precise and backwards compatible. Like so:

![image](https://github.com/directus/directus/assets/14810858/d46af388-60ff-4b69-afa2-c318ea32d3db)

If we add the new status we can give it a more concrete name too like `Unverified` for example

Reply:
I would have to take a closer look, but I think a dedicated (hidden) field for the email verification status could make sense. This could also be used, for example, if the user is already active and changes the email address, which would have to be verified again (but the user status should stay "active").